### PR TITLE
WEBGL_draw_instanced_base_vertex_base_instance

### PR DESCRIPTION
**Pull request by juj**
Add support for WEBGL_draw_instanced_base_vertex_base_instance WebGL 2 draft extension. 

Move webgl headers from private lib generation time to public headers directory.

Add support for -s GL_SUPPORT_SIMPLE_ENABLE_EXTENSIONS=0 flag to allow dropping code size dependency to all possible WebGL extensions.

### CODE REVIEW

File: src/library_html5.js [hunk 1]
```diff
@@ -2684,22 +2684,51 @@ var LibraryJSEvents = {
...
-    '_webgl_acquireVertexArrayObjectExtension', '_webgl_acquireDrawBuffersExtension'],
+    'emscripten_webgl_enable_ANGLE_instanced_arrays',
+    'emscripten_webgl_enable_OES_vertex_array_object',
+    'emscripten_webgl_enable_WEBGL_draw_buffers',
 #endif
+#if MAX_WEBGL_VERSION >= 2
+    'emscripten_webgl_enable_WEBGL_draw_instanced_base_vertex_base_instance',
+#endif
+    ],
+
```
Comment (kripken) [inline src/library_html5.js:2697]: no space between deps and functions

File: src/settings.js [hunk 2]
```diff
@@ -474,6 +474,18 @@ var GL_EXTENSIONS_IN_PREFIXED_FORMAT = 1;
...
+// If true, the function emscripten_webgl_enable_extension() can be called to
+// enable any WebGL extension. If false, to save code size,
+// emscripten_webgl_enable_extension() cannot be called to enable any of extensions
+// 'ANGLE_instanced_arrays', 'OES_vertex_array_object', 'WEBGL_draw_buffers' or
+// 'WEBGL_draw_instanced_base_vertex_base_instance', but the dedicated functions
+// emscripten_webgl_enable_*() found in html5.h are used to enable each of those
+// extensions. This way code size is increased only for the extensions that are
+// actually used.
+// N.B. if setting this to 0, GL_SUPPORT_AUTOMATIC_ENABLE_EXTENSIONS must be set
+// to zero as well.
```
Comment (kainino0x) [inline src/settings.js:486]: if not already done, assert somewhere if these flags conflict? (I think it can be done in emcc.py?)

**Comment by juj** [comment]
Updated with a fixup and a test.

**Comment by shrekshao** [comment]
Hi @juj, @kainino0x  Some general emscripten contributing questions (new to the development here...):

1. To use WEBGL_draw_instanced_base_vertex_base_instance, do we add in our C++ file:
```c++
#include <webgl/webgl2_ext.h>
//...
emscripten_webgl_enable_extension(ctx, "WEBGL_draw_instanced_base_vertex_base_instance");
//...
glDrawArraysInstancedBaseInstanceWEBGL(GL_TRIANGLES, 0, 3, 1, 0);
```
I'm getting
```
error: undefined symbol: glDrawArraysInstancedBaseInstanceWEBGL (referenced by top-level compiled C/C++ code)
...
warning: _glDrawArraysInstancedBaseInstanceWEBGL may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
```
Am I missing something (cmd flag etc.) here or am I calling the function wrong?
Also is emscripten going to export regular GL/GLES version glDrawArraysInstancedBaseInstance(EXT) function to using the webgl extension?


2. Is specialized extension enabling function like `emscripten_webgl_enable_WEBGL_draw_instanced_base_vertex_base_instance` only needed for extensions only lives in webgl1 or webgl2? (i.e. we just use `emscripten_webgl_enable_extensions` for regular extensions available in both webgl1/2 context?)

**Comment by kripken** [comment]
I think this is just in WebGL2? That's off by default, so you need to build with `emcc -s WEBGL2`.

(Not sure about the second question.)

**Comment by shrekshao** [comment]
Thanks for the reply.
I'm now running with the following config under `test_browser.py`. 
```python
@requires_graphics_hardware
  def test_webgl_multi_draw(self):
    self.btest('webgl_multi_draw_test.c', '0', args=['-lGL', '-s', 'OFFSCREEN_FRAMEBUFFER=1', '-DEXPLICIT_SWAP=1', '-s', 'FULL_ES3=1', '-s', 'MAX_WEBGL_VERSION=2', '-s', 'MIN_WEBGL_VERSION=2'])

```

Still getting the same issue `GLctx.dibvbi` being `null`. do I need any extra build flags to get the extension exported correctly?

**Comment by kripken** [comment]
I don't know enough about the WebGL side here - @kainino0x and @juj would be best. But if you know what the WebGL side should look like, you can maybe read the emitted JS and see what's missing, then read the `src/library_webgl*.js` code to see how to add it.

**Comment by kainino0x** [comment]
The extension is enabled and `dibvbi` is set when you call emscripten_webgl_enable_WEBGL_draw_instanced_base_vertex_base_instance():
https://github.com/emscripten-core/emscripten/blob/7ad70139bb0fcfa707959b6be1c7bce75188fea5/src/library_webgl2.js#L989

**Comment by shrekshao** [comment]
Faceplam... case solved. I was testing on Mac validating chrome and this extension is not exposed on this platform... Please ignore my noise.

**Pull request merged**